src/Raven.Server/Documents/Queries/AST/JavascriptCodeQueryVisitor.cs

```diff
@@ -95,9 +96,92 @@ public override void VisitCompoundWhereExpression(BinaryExpression @where)
 
         public override void VisitMethod(MethodExpression expr)
         {
+            // The regex usage here is just temporary, to get things working. Ideally, we want to implement the string functions in Jint
+            if (expr.Name.Value.Equals("startswith", StringComparison.OrdinalIgnoreCase) && expr.Arguments.Count == 2)
+            {
+                _sb.Append("new Regexp('^");
+                var param = expr.Arguments[1] as ValueExpression;
+                _sb.Append(Regex.Escape(param.Token.Value));
+                _sb.Append("').test('");
+                VisitExpression(expr.Arguments[0]);
+                
+                _sb.Append("')");
+                return;
+            }
+            
+            if (expr.Name.Value.Equals("endswith", StringComparison.OrdinalIgnoreCase) && expr.Arguments.Count == 2)
+            {
+                _sb.Append("new Regexp('");
+                var param = expr.Arguments[1] as ValueExpression;
+                _sb.Append(Regex.Escape(param.Token.Value));
+                _sb.Append("$').test('");
+                VisitExpression(expr.Arguments[0]);
+                
+                _sb.Append("')");
+                return;
+            }
+            
+            if (expr.Name.Value.Equals("intersect", StringComparison.OrdinalIgnoreCase) && expr.Arguments.Count >= 2)
+            {
+                _sb.Append("(");
+                for (var index = 0; index < expr.Arguments.Count; index++)
+                {
+                    var argument = expr.Arguments[index];
+                    
+                    VisitExpression(argument);
+                    
+                    if (index < expr.Arguments.Count - 1)
+                        _sb.Append(" && ");
+                }
+                
+                _sb.Append("')");
+                return;    
+            }
+            
+            if (expr.Name.Value.Equals("regex", StringComparison.OrdinalIgnoreCase) && expr.Arguments.Count == 2)
+            {
+                _sb.Append("new Regexp(' "); 
```

Comment:
Better to implement it ourselves, we need to make sure that it has the same beahvior as the where clause.

Reply:
it's the same .net Regex object in both of them, the differences are really minor
https://github.com/sebastienros/jint/blob/dev/Jint/Native/RegExp/RegExpInstance.cs
https://github.com/sebastienros/jint/blob/dev/Jint/Native/RegExp/RegExpPrototype.cs#L80
https://github.com/ravendb/ravendb/blob/v4.0/src/Raven.Server/Documents/Queries/Regex/RegexTermEnum.cs

On the other hand, we would probably like to perform some regex object caching, so maybe we still have to implement it ourselves...

Reply:
We want the caching and we want to get the timeout behavior

Reply:
you mean cache timeout?

Reply:
MatchTimeout